man/man1/pkcsep11_migrate.1.in
.TH PKCSEP11_MIGRATE 1 "August 2026" "@PACKAGE_VERSION@" "openCryptoki"
.SH NAME
pkcsep11_migrate \- utility to re-encrypt EP11 token keys to prepare a
change of master keys in the EP11 adapter

.SH SYNOPSIS
\fBpkcsep11_migrate\fP
[\fB-h\fP | \fB--help\fP]
\fB-slot\fP \fIslot-number\fP \fB-adapter\fP \fIadapter-ID\fP
\fB-domain\fP \fIdomain-ID\fP

.SH DESCRIPTION
In case of a master key change within an EP11 adapter, all key objects that are
wrapped with this master key must be re-wrapped or re-encrypted.
The \fBpkcsep11_migrate\fP utility takes all EP11 token-related key objects
that are wrapped with the EP11 adapter master key, decrypts each key object
with the current master key, and encrypts it with the new master key.

Notes:
.br
1. The new master key must be set and committed on the EP11 adapter via the
Trusted Key Entry (TKE) console before using this utility.
.br
2. While this tool is running, no process using the EP11 token should be running.
.br
3. Before using this tool, make a backup of the token objects in
   \fBep11tok/TOK_OBJ/\fP.
.br
4. After successful execution of the migrate utility and before (re)starting
   programs using the EP11 token, the new master key must be activated using
   the TKE.
.br
5. The utility prompts for the USER PIN of the EP11 token.

.SH "COMMAND SUMMARY"
.IP "\fB-slot\fP \fIslot-number\fP" 10
Specifies the token slot of the EP11 token. This option is required. The slot
number must be specified in decimal notation.
.IP "\fB-adapter\fP \fIadapter-ID\fP" 10
Specifies an EP11 adapter ID. This option is required.
(Refer to \fBlszcrypt\fP to get a list of installed crypto adapters.)
The adapter ID can be specified in hexadecimal (for example, \fB0x0A\fP) or
in decimal notation.
.IP "\fB-domain\fP \fIdomain-ID\fP" 10
Specifies the usage domain for the EP11 adapter. This option is required.
(Refer to \fBlszcrypt\fP to get a list of installed crypto domains.)
The domain ID can be specified in hexadecimal (for example, \fB0x0B\fP) or in
decimal notation.
.IP "\fB-h\fP, \fB--help\fP" 10
Show usage information.

.SH ENVIRONMENT
.IP "\fBPKCSLIB\fP" 10
Specifies the PKCS#11 library to load. If not set, \fBlibopencryptoki.so\fP
is used.
.IP "\fBOCK_EP11_LIBRARY\fP" 10
Specifies the EP11 host library to load. If not set, the utility tries
\fBlibep11.so.4\fP, \fBlibep11.so.3\fP, \fBlibep11.so.2\fP,
\fBlibep11.so.1\fP, and \fBlibep11.so\fP.

.SH SEE ALSO
.PD 0
.TP
\fBpkcsconf\fP(1),
.TP
\fBopencryptoki\fP(7),
.TP
\fBpkcsslotd\fP(8).
.PD


NEWS
VERSION 3.2 ADDITIONS:
 - A new tool to assist in administering the CCA token in opencryptoki.
   Currently it only assists in migrating CCA token's private token objects
   from version 2  to version 3. 
   Prior to opencryptoki version 3, the CCA token encrypted its private token
   objects with a secure key in hardware. In version 3, it encrypts its
   private token objects with a clear key in software. Version 2 private
   token objects will require migration to version 3, if they wished to be
   used in version 3.

VERSION 3.1 ADDITIONS:
 - New ep11 token. Please see doc/README.ep11_stdll for additional
   information about the ep11 token.

 - New pkcsep11_migrate utility to assist in migrating token objects
   stored on disk, when the hardware's masterkey changes.

VERSION 3 CHANGES:
 - New ICSF token. Please see doc/README.icsf_stdll for additional
   information and how to set up the ICSF token.

 - New pkcsicsf utility that is used to set up the ICSF token.
   See its man page and doc/README.icsf_stdll for additional info.

 - No longer required to run pkcs11_startup.

   pkcs11_startup and pkcs_slotd were shell scripts that created the
   config file, pk_config_data, which was read by pkcsslotd to get
   available slot information.

   The pk_config_data configuration file has been replaced with
   /etc/opencryptoki/opencryptoki.conf.

   In version 3, the pkcsslotd daemon reads opencryptoki.conf to get slot
   information. The opencryptoki.conf by default contains slot information
   for each token currently supported by opencryptoki with the exception of
   the ICSF token, which requires some initial setup.
   Please see man page for opencryptoki.conf for further information.

   Since pk_config_data is no longer required, pkcs11_startup and
   pkcs_slotd have been removed.

 - The pkcsslotd daemon uses a socket rather than shared memory to
   transfer slot information to the opencryptoki library.

   Some shared memory usage still exists in pkcsslotd.
   Perhaps in time, the remaining need for shared memory in
   pkcsslotd can also be removed.

 - Ensure that the pkcs11 group has been created and that root
   has been added to the group. The pkcs11_startup script used to 
   check and do this, but is now obsolete. 


NOTICEABLE CHANGES MADE PRIOR to VERSION 3:

 - Opencryptoki creates several new directories and lock files in
   /var/lock/opencryptoki directory. Each token creates and uses a 
   lockfile to protect data in shared memory.

 - RSA keys may be imported into the CCA token.

 - Opencryptoki contains systemd support. 
   Note: Ensure the opencryptoki lockfiles have been entered into tmpfiles.d
         to keep them persistant across reboots.



usr/sbin/pkcsep11_migrate/pkcsep11_migrate.c
/*
 * COPYRIGHT (c) International Business Machines Corp. 2013-2017
 *
 * This program is provided under the terms of the Common Public License,
 * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
 * software constitutes recipient's acceptance of CPL-1.0 terms which can be
 * found in the file LICENSE file or at
 * https://opensource.org/licenses/cpl1.0.php
 */

/* Reenryption of EP11 secure keys. The secure key is reencrypted at the card
 * by a new wrapping (still pending) key. Needed also for SPKIs of public
 * keys.
 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <memory.h>
#include <dlfcn.h>
#include <ctype.h>
#include <errno.h>
#include <openssl/crypto.h>

#define OCK_NO_EP11_DEFINES
#include "../../include/pkcs11types.h"
#include "../../lib/common/p11util.h"
#include "../../lib/ep11_stdll/ep11_func.h"
#include "pin_prompt.h"

#define EP11SHAREDLIB_NAME "OCK_EP11_LIBRARY"
#define EP11SHAREDLIB_V4 "libep11.so.4"
#define EP11SHAREDLIB_V3 "libep11.so.3"
#define EP11SHAREDLIB_V2 "libep11.so.2"
#define EP11SHAREDLIB_V1 "libep11.so.1"
#define EP11SHAREDLIB "libep11.so"
#define PKCS11_MAX_PIN_LEN 128

CK_FUNCTION_LIST *funcs;
CK_SLOT_ID SLOT_ID = -1;
CK_LONG adapter = -1;
CK_LONG domain = -1;
CK_OBJECT_HANDLE key_store[4096];

m_get_xcp_info_t _m_get_xcp_info;
m_admin_t _m_admin;
xcpa_cmdblock_t _xcpa_cmdblock;
xcpa_internal_rv_t _xcpa_internal_rv;
m_add_module_t _m_add_module;
m_rm_module_t _m_rm_module;

CK_VERSION lib_version;

typedef struct {
    short format;
    short length;
    short apqns[512];
} __attribute__ ((packed)) ep11_target_t;


#define BLOBSIZE         2048*4

static CK_RV ep11_reencrypt(target_t target, CK_BYTE *blob, CK_ULONG blob_len)
{
    CK_BYTE req[BLOBSIZE];
    CK_BYTE resp[BLOBSIZE];
    CK_LONG req_len;
    size_t resp_len;
    struct XCPadmresp rb;
    struct XCPadmresp lrb;
    CK_RV rc;

    memset(&rb, 0, sizeof(rb));
    memset(&lrb, 0, sizeof(lrb));

    rb.domain = domain;
    lrb.domain = domain;

    resp_len = BLOBSIZE;

    req_len = _xcpa_cmdblock(req, BLOBSIZE, XCP_ADM_REENCRYPT, &rb,
                             NULL, blob, blob_len);

    if (req_len < 0) {
        fprintf(stderr, "reencrypt cmd block construction failed\n");
        return -4;
    }

    rc = _m_admin(resp, &resp_len, NULL, 0, req, req_len, NULL, 0,
                  target);
    if (rc != CKR_OK || resp_len == 0) {
        fprintf(stderr, "reencryption failed: %lx %ld\n", rc, req_len);
        return -5;
    }

    if (_xcpa_internal_rv(resp, resp_len, &lrb, &rc) < 0) {
        fprintf(stderr, "reencryption response malformed: %lx\n", rc);
        return -6;
    }

    if (rc != 0) {
        fprintf(stderr, "reencryption failed: %lx\n", rc);
        return -7;
    }

    if (blob_len != lrb.pllen) {
        fprintf(stderr, "reencryption blob size changed: %lx %lx %lx %lx\n",
                blob_len, lrb.pllen, resp_len, req_len);
        return -8;
    }

    memcpy(blob, lrb.payload, blob_len);

    return 0;
}

static int reencrypt(CK_SESSION_HANDLE session, CK_ULONG obj,
                     CK_KEY_TYPE keytype, CK_BYTE *old, CK_ULONG old_len)
{
    ep11_target_t target_list;
    struct XCP_Module module;
    target_t target = XCP_TGT_INIT;
    CK_RV rc;
    CK_BYTE name[256];
    unsigned char blob[BLOBSIZE];
    CK_ULONG blob_len;

    CK_ATTRIBUTE opaque_template[] = {
        { CKA_IBM_OPAQUE, blob, BLOBSIZE }
    };

    CK_ATTRIBUTE name_template[] = {
        {CKA_LABEL, NULL_PTR, 0}
    };

    memset(name, 0, 256);

    /* print CKA_LABEL if it exists, only informational
       exist and size query */
    rc = funcs->C_GetAttributeValue(session, key_store[obj], name_template, 1);
    if (rc == CKR_OK && name_template[0].ulValueLen < 256) {
        name_template[0].pValue = name;
        /* knowing its size, after mem allocation, get the name value */
        rc = funcs->C_GetAttributeValue(session, key_store[obj], name_template,
                                        1);
    }

    if (rc != CKR_OK) {
        fprintf(stderr,
                "reencryption C_GetAttributeValue failed: obj %lx rc: %lx\n",
                obj, rc);
        return -1;
    }

    if (_m_add_module != NULL) {
        memset(&module, 0, sizeof(module));
        module.version = lib_version.major >= 3 ? XCP_MOD_VERSION_2
                                                : XCP_MOD_VERSION_1;
        module.flags = XCP_MFL_MODULE;
        module.module_nr = adapter;
        XCPTGTMASK_SET_DOM(module.domainmask, domain);
        rc = _m_add_module(&module, &target);
        if (rc != 0) {
            fprintf(stderr,
                    "reencryption m_add_module failed: rc: %lx\n",rc);
            return -2;
        }
    } else {
        /* Fall back to old target handling */
        memset(&target_list, 0, sizeof(ep11_target_t));
        target_list.length = 1;
        target_list.apqns[0] = adapter;
        target_list.apqns[1] = domain;
        target = (target_t)&target_list;
    }


    fprintf(stderr, "going to reencrpyt key %lx with blob len %lx: '%s'\n", obj,
            old_len, name);

    if (old_len > sizeof(blob)) {
        fprintf(stderr,
                "reencryption blob is too large: %lu\n", old_len);
        rc = -3;
        goto out;
    }

    blob_len = old_len;
    memcpy(blob, old, blob_len);

    if (keytype == CKK_AES_XTS) {
        /* An AES XTS key consists of 2 blobs concatenated to each other */
        rc = ep11_reencrypt(target, blob, blob_len / 2);
        if (rc != 0)
            goto out;

        rc = ep11_reencrypt(target, blob + blob_len / 2, blob_len / 2);
        if (rc != 0)
            goto out;
    } else {
        rc = ep11_reencrypt(target, blob, blob_len);
        if (rc != 0)
            goto out;
    }

    opaque_template[0].ulValueLen = blob_len;

    rc = funcs->C_SetAttributeValue(session, key_store[obj], opaque_template,
                                    1);
    if (rc != CKR_OK) {
        fprintf(stderr,
                "reencryption C_SetAttributeValue failed: obj %lx '%s' rc: %lx\n",
                obj, name, rc);
        rc = -9;
        goto out;
    }

    fprintf(stderr, "reencryption success obj: %lx '%s:\n", obj, name);

out:
    if (_m_add_module != NULL)
        _m_rm_module(&module, target);
    return rc;
}

static CK_RV get_ep11_library_version(CK_VERSION *lib_version)
{
    unsigned int host_version;
    CK_ULONG version_len = sizeof(host_version);
    CK_RV rc;

    rc = _m_get_xcp_info(&host_version, &version_len,
                         CK_IBM_XCPHQ_VERSION, 0, 0);
    if (rc != CKR_OK) {
        fprintf(stderr, "_m_get_xcp_info (HOST) failed: rc=0x%lx\n", rc);
        return rc;
    }
    lib_version->major = (host_version & 0x00FF0000) >> 16;
    /* Minor is 4 bits release number and 4 bits modification level */
    lib_version->minor = (host_version & 0x00000F00) >> 4 |
                                         (host_version & 0x0000000F);
    if ((host_version & 0x0000F000) != 0)
        lib_version->minor |= 0xF0;
    if ((host_version & 0x000000F0) != 0)
        lib_version->minor |= 0x0F;
    /*
     * EP11 host library < v2.0 returns an invalid version (i.e. 0x100). This
     * can safely be treated as version 1.0
     */
    if (lib_version->major == 0) {
        lib_version->major = 1;
        lib_version->minor = 0;
    }

    return CKR_OK;
}


static int check_card_status(void)
{
    CK_RV rc;
    ep11_target_t target_list;
    struct XCP_Module module;
    target_t target = XCP_TGT_INIT;
    CK_IBM_DOMAIN_INFO dinf;
    CK_ULONG dinf_len = sizeof(dinf);

    if (adapter == -1 || domain == -1) {
        fprintf(stderr, "adapter/domain specification missing.\n");
        return -1;
    }

    if (_m_add_module != NULL) {
        memset(&module, 0, sizeof(module));
        module.version = lib_version.major >= 3 ? XCP_MOD_VERSION_2
                                                : XCP_MOD_VERSION_1;
        module.flags = XCP_MFL_MODULE;
        module.module_nr = adapter;
        XCPTGTMASK_SET_DOM(module.domainmask, domain);
        rc = _m_add_module(&module, &target);
        if (rc != 0)
            return CKR_FUNCTION_FAILED;
    } else {
        /* Fall back to old target handling */
        memset(&target_list, 0, sizeof(ep11_target_t));
        target_list.length = 1;
        target_list.apqns[0] = adapter;
        target_list.apqns[1] = domain;
        target = (target_t)&target_list;
    }

    rc = _m_get_xcp_info((CK_VOID_PTR) &dinf, &dinf_len,
                         CK_IBM_XCPQ_DOMAIN, 0, target);

    if (rc != CKR_OK) {
        fprintf(stderr, "m_get_xcp_info rc 0x%lx, valid apapter/domain "
                "0x%02lx/%ld?.\n", rc, adapter, domain);
        rc = -1;
        goto out;
    }

    if (CK_IBM_DOM_COMMITTED_NWK & dinf.flags) {
        fprintf(stderr, "Card ID 0x%02lx, domain ID %ld has committed "
                "pending(next) WK\n", adapter, domain);
    } else {
        fprintf(stderr,
                "Card ID 0x%02lx, domain ID %ld has no committed pending WK\n",
                adapter, domain);
        rc = -1;
        goto out;
    }

out:
    if (_m_add_module != NULL)
        _m_rm_module(&module, target);

    return rc;
}

static int get_user_pin(CK_BYTE *dest)
{
    int ret = -1;
    const char *userpin = NULL;
    char *buf_user = NULL;
    size_t userpinlen;

    userpin = pin_prompt(&buf_user, "Enter the USER PIN: ");
    if (!userpin) {
        fprintf(stderr, "Could not get USER PIN.\n");
        goto out;
    }
    userpinlen = strlen(userpin);

    if (userpinlen > PKCS11_MAX_PIN_LEN) {
        fprintf(stderr, "The USER PIN must be less than %d chars in length.\n",
                (int) PKCS11_MAX_PIN_LEN);
        goto out;
    }

    memcpy(dest, userpin, userpinlen + 1);
    ret = 0;
out:
    pin_free(&buf_user);
    return ret;
}

static int do_GetFunctionList(void)
{
    CK_RV rc;
    CK_RV (*func_list)(CK_FUNCTION_LIST_PTR_PTR ppFunctionList) = NULL;
    void *d;
    char *evar;
    char *evar_default = "libopencryptoki.so";

    evar = secure_getenv("PKCSLIB");
    if (evar == NULL) {
        evar = evar_default;
    }

    d = dlopen(evar, RTLD_NOW);
    if (d == NULL) {
        return 0;
    }

    *(void **)(&func_list) = dlsym(d, "C_GetFunctionList");
    if (func_list == NULL) {
        return 0;
    }
    rc = func_list(&funcs);

    if (rc != CKR_OK) {
        return 0;
    }

    return 1;

}

static void usage(char *fct)
{
    printf("usage:  %s [-slot <num>] [-adapter <num>] [-domain <num>] [-h]\n\n",
           fct);
    return;
}

static int do_ParseArgs(int argc, char **argv)
{
    int i;
    unsigned long val;
    char *endptr;

    if (argc <= 1) {
        printf("No Arguments given. "
               "For help use the '--help' or '-h' option.\n");
        return -1;
    }

    for (i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
            usage(argv[0]);
            return 0;
        } else if (strcmp(argv[i], "-slot") == 0) {
            if (i + 1 >= argc) {
                printf("Slot parameter is missing!\n");
                return -1;
            }
            errno = 0;
            val = strtoul(argv[i + 1], &endptr, 10);
            if (errno != 0 || endptr == argv[i + 1] || *endptr != '\0'
                || argv[i + 1][0] == '-') {
                printf("Slot parameter is not a valid number!\n");
                return -1;
            }
            SLOT_ID = (CK_SLOT_ID) val;
            i++;
        } else if (strcmp(argv[i], "-adapter") == 0) {
            if (i + 1 >= argc) {
                printf("Adapter parameter is missing!\n");
                return -1;
            }
            errno = 0;
            val = strtoul(argv[i + 1], &endptr,
                          strncasecmp(argv[i + 1], "0x", 2) == 0 ? 16 : 10);
            if (errno != 0 || endptr == argv[i + 1] || *endptr != '\0'
                || argv[i + 1][0] == '-') {
                printf("Adapter parameter is not a valid number!\n");
                return -1;
            }
            adapter = (CK_LONG) val;
            i++;
        } else if (strcmp(argv[i], "-domain") == 0) {
            if (i + 1 >= argc) {
                printf("Domain parameter is missing!\n");
                return -1;
            }
            errno = 0;
            val = strtoul(argv[i + 1], &endptr,
                          strncasecmp(argv[i + 1], "0x", 2) == 0 ? 16 : 10);
            if (errno != 0 || endptr == argv[i + 1] || *endptr != '\0'
                || argv[i + 1][0] == '-') {
                printf("Domain parameter is not a valid number!\n");
                return -1;
            }
            domain = (CK_LONG) val;
            i++;
        } else {
            printf("Invalid argument passed as option: %s\n", argv[i]);
            usage(argv[0]);
            return -1;
        }
    }
    if (SLOT_ID == (CK_SLOT_ID)(-1)) {
        printf("Slot-ID not set!\n");
        return -1;
    }
    if (adapter == -1) {
        printf("Adapter-ID not set!\n");
        return -1;
    }
    if (domain == -1) {
        printf("Domain-ID not set!\n");
        return -1;
    }

    return 1;
}

#ifdef EP11_HSMSIM
#define DLOPEN_FLAGS        RTLD_GLOBAL | RTLD_NOW | RTLD_DEEPBIND
#else
#define DLOPEN_FLAGS        RTLD_GLOBAL | RTLD_NOW
#endif

static void *ep11_load_host_lib(void)
{
    void *lib_ep11;
    char *ep11_lib_name;
    char *errstr;

    ep11_lib_name = secure_getenv(EP11SHAREDLIB_NAME);
    if (ep11_lib_name != NULL) {
        lib_ep11 = dlopen(ep11_lib_name, DLOPEN_FLAGS);

        if (lib_ep11 == NULL) {
            errstr = dlerror();
            fprintf(stderr, "Error loading shared library '%s' [%s]\n",
                    ep11_lib_name, errstr);
            return NULL;
        }
        return lib_ep11;
    }

    ep11_lib_name = EP11SHAREDLIB_V4;
    lib_ep11 = dlopen(ep11_lib_name, DLOPEN_FLAGS);

    if (lib_ep11 == NULL) {
        /* Try version 3 instead */
        ep11_lib_name = EP11SHAREDLIB_V3;
        lib_ep11 = dlopen(ep11_lib_name, DLOPEN_FLAGS);
    }

    if (lib_ep11 == NULL) {
        /* Try version 2 instead */
        ep11_lib_name = EP11SHAREDLIB_V2;
        lib_ep11 = dlopen(ep11_lib_name, DLOPEN_FLAGS);
    }

    if (lib_ep11 == NULL) {
        /* Try version 1 instead */
        ep11_lib_name = EP11SHAREDLIB_V1;
        lib_ep11 = dlopen(ep11_lib_name, DLOPEN_FLAGS);
    }

    if (lib_ep11 == NULL) {
        /* Try unversioned library instead */
        ep11_lib_name = EP11SHAREDLIB;
        lib_ep11 = dlopen(ep11_lib_name, DLOPEN_FLAGS);
    }

    if (lib_ep11 == NULL) {
        errstr = dlerror();
        fprintf(stderr, "Error loading shared library '%s[.4|.3|.2|.1]' [%s]\n",
                EP11SHAREDLIB, errstr);
        return NULL;
    }

    return lib_ep11;
}

int main(int argc, char **argv)
{
    int rc;
    void *lib_ep11;
    CK_C_INITIALIZE_ARGS cinit_args;
    CK_BYTE user_pin[PKCS11_MAX_PIN_LEN + 1];
    CK_FLAGS flags;
    CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
    CK_ULONG obj;
    CK_ULONG user_pin_len;
    CK_ULONG keys_found = 0;
    CK_BBOOL logged_in = CK_FALSE;

    rc = do_ParseArgs(argc, argv);
    if (rc != 1) {
        return rc;
    }

    /* dynamically load in the ep11 shared library */
    lib_ep11 = ep11_load_host_lib();
    if (!lib_ep11)
        return CKR_FUNCTION_FAILED;

    *(void **)(&_xcpa_cmdblock) = dlsym(lib_ep11, "xcpa_cmdblock");
    if (_xcpa_cmdblock == NULL)
        *(void **)(&_xcpa_cmdblock) = dlsym(lib_ep11, "ep11a_cmdblock");
    *(void **)(&_m_admin) = dlsym(lib_ep11, "m_admin");
    *(void **)(&_xcpa_internal_rv) = dlsym(lib_ep11, "xcpa_internal_rv");
    if (_xcpa_internal_rv == NULL)
        *(void **)(&_xcpa_internal_rv) = dlsym(lib_ep11, "ep11a_internal_rv");
    *(void **)(&_m_get_xcp_info) = dlsym(lib_ep11, "m_get_xcp_info");
    if (_m_get_xcp_info == NULL)
        *(void **)(&_m_get_xcp_info) = dlsym(lib_ep11, "m_get_ep11_info");

    if (!_m_get_xcp_info || !_xcpa_cmdblock ||
        !_m_admin || !_xcpa_internal_rv) {
        fprintf(stderr, "ERROR getting function pointer from shared lib '%s'",
                EP11SHAREDLIB);
        return CKR_FUNCTION_FAILED;
    }

    /*
     * The following are only available since EP11 host library version 2.
     * Ignore if they fail to load, the code will fall back to the old target
     * handling in this case.
     */
    *(void **)(&_m_add_module) = dlsym(lib_ep11, "m_add_module");
    *(void **)(&_m_rm_module) = dlsym(lib_ep11, "m_rm_module");
    if (_m_add_module == NULL || _m_rm_module == NULL) {
        _m_add_module = NULL;
        _m_rm_module = NULL;
    }

    printf("Using slot #%lu...\n\n", SLOT_ID);

    rc = do_GetFunctionList();
    if (!rc) {
        fprintf(stderr, "ERROR do_GetFunctionList() Failed, rx = 0x%0x\n", rc);
        return rc;
    }

    memset(&cinit_args, 0x0, sizeof(cinit_args));
    cinit_args.flags = CKF_OS_LOCKING_OK;

    funcs->C_Initialize(&cinit_args);
    {
        CK_SESSION_HANDLE hsess = 0;
        rc = funcs->C_GetFunctionStatus(hsess);
        if (rc != CKR_FUNCTION_NOT_PARALLEL) {
            return rc;
        }

        rc = funcs->C_CancelFunction(hsess);
        if (rc != CKR_FUNCTION_NOT_PARALLEL) {
            return rc;
        }
    }

    flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
    rc = funcs->C_OpenSession(SLOT_ID, flags, NULL, NULL, &session);
    if (rc != CKR_OK) {
        fprintf(stderr, "C_OpenSession() rc = 0x%02x [%s]\n", rc,
                p11_get_ckr(rc));
        session = CK_INVALID_HANDLE;
        return rc;
    }

    if (get_user_pin(user_pin)) {
        fprintf(stderr, "get_user_pin() failed\n");
        rc = CKR_FUNCTION_FAILED;
        goto out;
    }

    user_pin_len = (CK_ULONG) strlen((char *) user_pin);
    rc = funcs->C_Login(session, CKU_USER, user_pin, user_pin_len);
    OPENSSL_cleanse(user_pin, sizeof(user_pin));
    if (rc != CKR_OK) {
        fprintf(stderr, "C_Login() rc = 0x%02x [%s]\n", rc, p11_get_ckr(rc));
        goto out;
    }
    logged_in = CK_TRUE;

    rc = get_ep11_library_version(&lib_version);
    if (rc != CKR_OK)
        goto out;

    if (check_card_status() != 0) {
        rc = 1;
        goto out;
    }

    /* find all objects */
    rc = funcs->C_FindObjectsInit(session, NULL, 0);
    if (rc != CKR_OK) {
        fprintf(stderr, "C_FindObjectsInit() rc = 0x%02x [%s]\n", rc,
                p11_get_ckr(rc));
        goto out;
    }

    do {
        rc = funcs->C_FindObjects(session, key_store, 4096, &keys_found);

        if (rc != CKR_OK) {
            fprintf(stderr, "C_FindObjects() rc = 0x%02x [%s]\n", rc,
                    p11_get_ckr(rc));
            goto out;
        }

        for (obj = 0; obj < keys_found; obj++) {
            CK_ATTRIBUTE opaque_template[] = {
                {CKA_IBM_OPAQUE, NULL_PTR, 0}
            };

            CK_KEY_TYPE keytype;
            CK_ATTRIBUTE key_type_template[] = {
                {CKA_KEY_TYPE, &keytype, sizeof(CK_KEY_TYPE)}
            };

            CK_BYTE *old_blob;

            /* only for keys */
            rc = funcs->C_GetAttributeValue(session, key_store[obj],
                                            key_type_template, 1);
            if (rc != CKR_OK)
                continue;

            /* exist and size query CKA_IBM_QPAQUE */
            rc = funcs->C_GetAttributeValue(session, key_store[obj],
                                            opaque_template, 1);
            if (rc == CKR_OK) {
                old_blob = malloc(opaque_template[0].ulValueLen);
                if (old_blob == NULL) {
                    fprintf(stderr, "malloc failed for opaque blob\n");
                    rc = -1;
                    goto out;
                }
                opaque_template[0].pValue = old_blob;
                /* get the blob after knowing its size */
                rc = funcs->C_GetAttributeValue(session, key_store[obj],
                                                opaque_template, 1);

                if (rc != CKR_OK) {
                    fprintf(stderr, "second C_GetAttributeValue failed "
                            "rc = 0x%02x [%s]\n", rc, p11_get_ckr(rc));
                    free(old_blob);
                    goto out;
                }
                if (reencrypt(session, obj, keytype,
                              (CK_BYTE *) opaque_template[0].pValue,
                              opaque_template[0].ulValueLen) != 0) {
                    /* reencrypt failed */
                    free(old_blob);
                    rc = -1;
                    goto out;
                }
                free(old_blob);
            }
        }
    }
    /* next 4096 objects */
    while (keys_found != 0);

    funcs->C_FindObjectsFinal(session);
    fprintf(stderr, "all keys successfully reencrypted\n");
    rc = CKR_OK;

out:
    if (logged_in)
        funcs->C_Logout(session);
    if (session != CK_INVALID_HANDLE)
        funcs->C_CloseAllSessions(SLOT_ID);

    return rc;
}
